Make MPI routines more restrictive and improve their docs

Thoemi09 opened the pull request:
Improve the MPI routines by making the requirements more restrictive and add some more checks.

Wentzell commented:
Thank you @Thoemi09 for these changes, can you split the doc changes the source code changes for easier review?

Thoemi09 commented:
Sure, I can do that

Thoemi09 committed:
Update mpi_reduce and add more tests

- change the type stored in the lazy object
- change the shape of the resulting ArrayInitializer object
- make the checks more consistent
- simplify by using the range functions from mpi

Thoemi09 committed:
Update mpi_scatter and add more tests

- change the type stored in the lazy object
- make the checks more consistent
- restrict to C order arrays/views
- simplify by using the range functions from mpi

Thoemi09 committed:
Update the docs of the MPI interface

Wentzell reviewed (approved):
Thank you @Thoemi09 for these improvements!
This has now been merged with minor modifications into unstable
